engines/gnap/gamesys.cpp

```diff
@@ -0,0 +1,1278 @@
+	const int height = spriteResource->_height;
+	_vm->_spriteCache->release(resourceId);
+	return height;
+}
+
+Graphics::Surface *GameSys::loadBitmap(int resourceId) {
+	debugC(kDebugBasic, "GameSys::loadBitmap() resourceId: %08X", resourceId);
+	if (_vm->_dat->getResourceType(resourceId) != 1)
+		return nullptr;
+	byte *resourceData = _vm->_dat->loadResource(resourceId);
+	uint32 resourceSize = _vm->_dat->getResourceSize(resourceId);
+	Common::MemoryReadStream stream(resourceData, resourceSize, DisposeAfterUse::NO);
+	Graphics::Surface *bmpSurface;
+	Image::BitmapDecoder bmp;
+	if (!bmp.loadStream(stream))
+		error("GameSys::loadBitmap() Could not load bitmap resource %08X", resourceId);
+	bmpSurface = bmp.getSurface()->convertTo(_vm->_system->getScreenFormat());
+	delete[] resourceData;
+	return bmpSurface;
+}
+
+void GameSys::drawBitmap(int resourceId) {
+	Graphics::Surface *bmpSurface = loadBitmap(resourceId);
+	if (!bmpSurface || !_backgroundSurface) {
+		debugC(kDebugBasic, "GameSys::drawBitmap() Error loading the bitmap");
+		return;
+	}
+	if (bmpSurface->format != _backgroundSurface->format ||
+		bmpSurface->w != _backgroundSurface->w || bmpSurface->h != _backgroundSurface->h) {
+		debugC(kDebugBasic, "GameSys::drawBitmap() Different bitmap properties than current background");
+	} else {
+		byte *src = (byte*)bmpSurface->getPixels();
+		byte *dst = (byte*)_backgroundSurface->getPixels();
+		const int pitch = bmpSurface->pitch;
+		int height = bmpSurface->h;
+		while (height--) {
+			memcpy(dst, src, pitch);
+			src += pitch;
+			dst += pitch;
+		}
+	}
+	bmpSurface->free();
+	delete bmpSurface;
+
+	insertDirtyRect(Common::Rect(0, 0, 800, 600));
+}
+
+Sequence *GameSys::seqFind(int sequenceId, int id, int *outIndex) {
+	for (uint i = 0; i < _seqItems.size(); ++i)
+		if (_seqItems[i]._sequenceId == sequenceId && _seqItems[i]._id == id) {
+			if (outIndex)
+				*outIndex = i;
+			return &_seqItems[i];
+		}
+	return nullptr;
+}
+
+int GameSys::seqLocateGfx(int sequenceId, int id, int *outGfxIndex) {
+	for (int i = 0; i < _gfxItemsCount; ++i) {
+		GfxItem *gfxItem = &_gfxItems[i];
+		if (gfxItem->_sequenceId == sequenceId && gfxItem->_id == id) {
+			if (outGfxIndex)
+				*outGfxIndex = i;
+			return gfxItem->_sequenceId;
+		}
+		if (gfxItem->_id > id) {
+			if (outGfxIndex)
+				*outGfxIndex = i;
+			return 0;
+		}
+	}
+	if (outGfxIndex)
+		*outGfxIndex = _gfxItemsCount;
+	return 0;
+}
+
+void GameSys::seqInsertGfx(int index, int duration) {
+	Sequence *seqItem = &_seqItems[index];
+	SequenceResource *sequenceResource = _vm->_sequenceCache->get(seqItem->_sequenceId);
+
+	if (sequenceResource->_animationsCount > 50 - _gfxItemsCount)
+		return;
+
+	int gfxIndex;
+	seqLocateGfx(seqItem->_sequenceId, seqItem->_id, &gfxIndex);
+
+	if (gfxIndex != _gfxItemsCount)
+		memcpy(&_gfxItems[gfxIndex + sequenceResource->_animationsCount],	&_gfxItems[gfxIndex], sizeof(GfxItem) * (_gfxItemsCount - gfxIndex));
```

Comment:
Valgrind complains about overlapping memory areas here, which is undefined, and to me it looks like that will always be the case so we should probably use memmove() here.